### PR TITLE
Features/ontouml2gufo

### PR DESCRIPTION
**Pull request by LucasBassetti**
Features included in this PR:

- [x] `owl:imports` for `gufo` ontology
- [x]  `owl:AllDisjointClasses` for all `kind`, `quantityKind`, `collectiveKind`, `relatorKind`, `modeKind` and `qualityKind`
- [x] Transformation for all class stereotypes
- [x] New ModelManager method: `getRelations`
- [x] Add Disjoint and Complete properties of generalization sets
- [x] Add snapshot tests

### CODE REVIEW

File: src/libs/ontuml2gufo/index.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+/**
+ * Utility class for transform OntoUML models in OWL using the gUFO ontology
+ *
+ * @author Claudenir Fonseca
+ * @author Lucas Bassetti
+ */
+export class OntoUML2GUFO {
+  private model: ModelManager;
+
+  constructor(model: ModelManager) {
```
Comment (claudenirmf) [inline src/libs/ontuml2gufo/index.ts]: You don't need the `ModelManager` here, you only need the `IPackage` that is a root package. The only thing is that the `IPackage` you use must first be updated by a model manager, for instance:

```typescript
const modelManager = new ModelManager(require('serialized-ontouml-model.json'));
const toGUFO = new OntoUML2GUFO(modelManager.rootPackage);
```

File: src/libs/ontuml2gufo/index.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+const { namedNode, literal, quad } = DataFactory;
+
+/**
+ * Utility class for transform OntoUML models in OWL using the gUFO ontology
+ *
+ * @author Claudenir Fonseca
+ * @author Lucas Bassetti
+ */
+export class OntoUML2GUFO {
+  private model: ModelManager;
```
Comment (claudenirmf) [inline src/libs/ontuml2gufo/index.ts]: I understand that using `private` is a good practice, but is there a problem if we leave this a public? Does it make any difference for Javascript projects importing this module?

File: src/libs/ontuml2gufo/index.ts (hunk 1 above)
Comment (LucasBassetti) [reply src/libs/ontuml2gufo/index.ts]: Should I replace the `model` variable to `pkg`?

File: src/constants/index.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  LITERAL_TYPE = 'Literal',
+}
+
+export enum AggregationKind {
+  AGGREGATIONKIND_NONE = 'NONE',
+  AGGREGATIONKIND_SHARED = 'SHARED',
+  AGGREGATIONKIND_COMPOSITE = 'COMPOSITE',
+}
+
+export enum ClassStereotype {
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply src/constants/index.ts]: No idea, we can keep that. If we have some problem in the future we can change for constant, like so:

```
export const OntoUMLType = {
  PACKAGE_TYPE: 'Package',
  CLASS_TYPE: 'Class',
  ...
}
```

File: src/home.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+const packages = require('./../test_models/others/packages.json');
+const Package = require('./../dist/libs/model').Package;
+// const ModelManager = require('./../dist/libs/model/').ModelManager;
+// const manager = new ModelManager(packages);
+
+var pkg = new Package();
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply src/home.ts]: ok

File: src/libs/ontuml2gufo/index.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+} from './class_stereotype_functions';
+
+const N3 = require('n3');
+const { DataFactory } = N3;
+const { namedNode, literal, quad } = DataFactory;
+
+/**
+ * Utility class for transform OntoUML models in OWL using the gUFO ontology
+ *
+ * @author Claudenir Fonseca
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply src/libs/ontuml2gufo/index.ts]: But you are reviewing 😃

File: src/libs/ontuml2gufo/index.ts (hunk 2 above)
Comment (LucasBassetti) [reply src/libs/ontuml2gufo/index.ts]: I will change this

File: src/libs/model/functions.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    if (!descendents.includes(child)) {
+      descendents = [...child.getDescendents(descendents)];
+      descendents.push(child);
+    }
+  });
+
+  return descendents;
+}
+
+function getRelations(): IRelation[] {
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply src/libs/model/functions.ts]: I didn't understand well but I agree xD

File: src/libs/ontuml2gufo/index.ts (hunk 1 above)
Comment (LucasBassetti) [reply src/libs/ontuml2gufo/index.ts]: I believe is better keep `model` in the `new` and change just internally. Like:

`const toGUFO = new OntoUML2GUFO(modelManager);`

```
export class OntoUML2GUFO {
  pkg: IPackage;

  constructor(model: ModelManager) {
    this.pkg = model.rootPackage;
  }

  ...
}
```

File: src/constants/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (claudenirmf) [reply src/constants/index.ts]: Ok

File: src/libs/ontuml2gufo/index.ts (hunk 1 above)
Comment (claudenirmf) [reply src/libs/ontuml2gufo/index.ts]: Ok, but I like to refer to the "root package" as "model". Even though it has the same features as any `IPackage` object, the concept of model is very traditional and conveys the clarity of having the whole thing in one entity that you can manipulate (e.g., transform from one language to another).

Comment (LucasBassetti) [reply src/libs/ontuml2gufo/index.ts]: Done

File: __tests__/libs/ontuml2gufo/classes.test.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { ModelManager } from '@libs/model';
+import { OntoUML2GUFO } from '@libs/ontuml2gufo';
+import { example, modePattern1, modePattern2 } from '@test-models/valids';
+
+it('Transform OntoUML model to gUFO', async () => {
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply __tests__/libs/ontuml2gufo/classes.test.ts]: Tests were improved

File: __tests__/libs/ontuml2gufo/classes.test.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+  expect(result.includes('owl:imports')).toBe(true);
+});
+
+it('Check OntoUML mode pattern 1 - IntrinsicMode', async () => {
+  const modelManager = new ModelManager(modePattern1);
+  const service = new OntoUML2GUFO(modelManager);
+  const result = await service.transformOntoUML2GUFO('https://example.com');
+
+  expect(result.includes('gufo:IntrinsicMode')).toBe(true);
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply __tests__/libs/ontuml2gufo/classes.test.ts]: Tests were improved

File: __tests__/libs/ontuml2gufo/classes.test.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+});
+
+it('Check OntoUML mode pattern 2 - ExtrinsicMode', async () => {
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply __tests__/libs/ontuml2gufo/classes.test.ts]: Tests were improved

File: __tests__/libs/ontuml2gufo/classes.test.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { ModelManager } from '@libs/model';
+import { OntoUML2GUFO } from '@libs/ontuml2gufo';
+import { example, modePattern1, modePattern2 } from '@test-models/valids';
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply __tests__/libs/ontuml2gufo/classes.test.ts]: Tests were improved

File: __tests__/libs/ontuml2gufo/classes.test.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import { ModelManager } from '@libs/model';
```
[the remark this answers is not in the document]
Comment (LucasBassetti) [reply __tests__/libs/ontuml2gufo/classes.test.ts]: Tests were improved

File: src/libs/ontuml2gufo/class_functions.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    ClassStereotype.COLLECTIVE,
+    ClassStereotype.RELATOR,
+    ClassStereotype.MODE,
+    ClassStereotype.QUALITY,
+  ];
+
+  for (let i = 0; i < disjointStereotypes.length; i += 1) {
+    const stereotype = disjointStereotypes[i];
+    const stereotypeClasses = classes
+      .filter(({ stereotypes }: IClass) => stereotypes[0] === stereotype)
```
Comment (tgoprince) [inline src/libs/ontuml2gufo/class_functions.ts]: This breaks if a class has no stereotypes.

**Comment by tgoprince** [comment]
I just commited two models as extra tests for you. 

One is the alpinebits ontology and the other the istandard ontology.

I noticed that on the istarndard ontology, the .ttl file we generate cannot be imported on protege. 
I think one of the issues it that the ids of some elements in the json model have . in their ids:

```json
{
   "type": "Class",
   "id": "Pg.JFHaGAqAe8BSn"
}
```

**Pull request merged**